File: server/heartbeat/triggers.js
// The event-driven half of Heartbeat — reacts to things HAPPENING (a job
// newly needing the owner's decision) rather than waiting for the next
// scheduled poll, while feeding the exact same pipeline (engine.js's
// routeFinding()) the tick uses. One pipeline, two entry points — never a
// second one.
//
// Subscribes to jobs/job-events.js's existing `jobEvents` EventEmitter — a
// SECOND listener alongside orchestrator.js's own and improvement/cycle.js's
// own; this emitter already supports multiple subscribers by design (see
// job-store.js's/orchestrator.js's own header comments on the two-channel
// pattern this mirrors).
//
// Deliberately does NOT react to every status transition — 'done'/'failed'/
// 'cancelled' already get a real addNotification() directly from
// worker.js/orchestrator.js (Jobs' own existing ambient channel); this only
// reacts to 'awaiting_decision', the one transition that means a fresh
// Tier 1 outbox row was very likely just created and is otherwise sitting
// silent until the user happens to start a conversation — exactly the gap
// jobs-source.js's own polling path exists to close, just without waiting
// up to that source's own check interval.

import { jobEvents } from '../jobs/job-events.js';
import { source as jobsSource } from './sources/jobs-source.js';
import { routeFinding } from './engine.js';
import * as scheduleStore from './schedule-store.js';

let subscribed = false;

async function onJobStatus({ jobId, status }) {
  if (status !== 'awaiting_decision') return;
  try {
    // Ensures a schedule row exists (harmless no-op if one already does)
    // and persists whatever checkState this check produces — without this,
    // a job's own dedup memory (see jobs-source.js's own header comment on
    // why it exists) would never actually get written by the TRIGGER path,
    // only by the regular tick, letting one redundant re-fire slip through
    // ~CHECK_INTERVAL_MS later even though nothing had changed.
    scheduleStore.upsertItem(jobsSource.id, jobId, jobsSource.defaultIntervalMs);
    const { finding, checkState } = await jobsSource.check(jobId);
    scheduleStore.markDone(`${jobsSource.id}:${jobId}`, checkState !== undefined ? { checkState } : undefined);
    if (finding) await routeFinding('jobs', jobId, finding);
  } catch (err) {
    console.error(`[heartbeat] trigger check failed for job "${jobId}":`, err);
  }
}

/** Idempotent — a second call is a no-op, same guard shape as orchestrator.js's own startOrchestrator(). Call once from heartbeat/index.js. */
export function startTriggers() {
  if (subscribed) return;
  jobEvents.on('status', (payload) => {
    onJobStatus(payload).catch((err) => console.error('[heartbeat] trigger handler failed:', err));
  });
  subscribed = true;
}

File: backend/jarvis/heartbeat/triggers.py
"""Reacting to something happening, rather than waiting for the next poll.

Feeds the SAME `route_finding()` the tick does — one pipeline, two entry points.
A second pipeline would mean two places deciding whether to interrupt someone.

Subscribes to the event bus rather than being called by the jobs code: a job
already publishes when it parks, and having the jobs subsystem know that a
heartbeat exists would be the coupling the bus was built to remove.

Only `awaiting_decision` is reacted to. Completions and failures already have
their own notification the moment they happen; this is for the one transition
that otherwise sits silent.
"""

from __future__ import annotations

import logging
from typing import Callable

from ..events import EventType, bus as default_bus
from ..events.bus import Event, EventBus

logger = logging.getLogger(__name__)

_unsubscribes: list[Callable[[], None]] = []


def _on_job_updated(event: Event) -> None:
    if event.payload.get("status") != "awaiting_decision":
        return
    job_id = event.payload.get("id")
    if not job_id:
        return

    from . import schedule_store
    from .engine import route_finding
    from .sources.jobs_source import source as jobs_source

    try:
        # Ensure a schedule row exists and persist whatever memory this check
        # produces. Without that, the trigger path would never write the dedup
        # state, and the next ordinary tick would report the same ask again.
        schedule_store.upsert_item(jobs_source.id, job_id, jobs_source.default_interval_ms)
        outcome = jobs_source.check(job_id) or {}
        if "checkState" in outcome:
            schedule_store.mark_done(f"{jobs_source.id}:{job_id}", outcome["checkState"])
        else:
            schedule_store.mark_done(f"{jobs_source.id}:{job_id}")
        if outcome.get("finding"):
            route_finding(jobs_source.id, job_id, outcome["finding"])
    except Exception:  # noqa: BLE001 — a trigger must never break the thing that fired it
        logger.exception("the trigger check for job %s failed", job_id)


def start_triggers(event_bus: EventBus | None = None) -> None:
    """Idempotent: a second call replaces the first rather than doubling it."""
    stop_triggers()
    ebus = event_bus or default_bus
    _unsubscribes.append(ebus.subscribe(EventType.JOB_UPDATED, _on_job_updated))


def stop_triggers() -> None:
    while _unsubscribes:
        try:
            _unsubscribes.pop()()
        except Exception:  # noqa: BLE001
            logger.warning("could not detach a heartbeat trigger")
